docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,289 @@
+
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+[PlantUML](http://plantuml.com) is the preferred tool to generate diagrams,
+place your `.plantuml` files under `images/` and run `make diagrams` from the docs folder.
+
+### User Stories
+#### Story 1
+(pre-delete) As an operator, I would like to have the ability to perform
+different actions between the time a machine is marked deleted in the api and
+the time the machine is deleted from the cloud.
+
+One use-case is when replacing a control plane machine, ensure a new control
+plane machine has been successfully created and joined to the cluster before
+removing the instance for the deleted machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine-controller.  This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+#### Story 3
+(pre-drain) As an operator, when I am deleting a control plane machine for
+maintenance reasons, I want to ensure my existing control plane machine is not
+drained until after my replacement comes online.  This will prevent protracted
+periods of a missing control plane member if the replacement machine cannot be
+created in a timely manner.
+
+
+### Implementation Details/Notes/Constraints
+
+For each defined lifecycle point, one or more hooks may be applied as an annotation to the machine object.  These annotations will pause reconciliation of a machine object until all hooks are resolved for that lifecycle point.  The hooks should be managed by an Owning Controller or other external application, or manually created and removed by an administrator.
+
+#### Lifecycle Points
+##### pre-drain
+Hooks defined at this point will prevent the machine-controller from draining a node after the machine-object has been marked for deletion until the hooks are removed.
+##### pre-term
+Hooks defined at this point will prevent the machine-controller from
+removing/terminating the instance in the cloud provider until the hooks are
+removed.
+
+
+#### Annotation Form
+```
+<lifecycle-point>.hook.machine.cluster-api.x-k8s.io/<hook-name>: <owner/creator>
+```
+
+##### lifecycle-point
+This is the point in the lifecycle of reconciling a machine the annotation will have effect and pause the machine-controller.
+
+##### hook-name
+Each hook should have a unique and descriptive name that describes in 1-3 words what the intent/reason for the hook is.  Each hook name should be unique and managed by a single entity.
```

Comment:
```suggestion
Each hook should have a unique and descriptive name that describes in 1-3 words what the intent/reason for the hook is.
```
I think we can have better guarantee of name uniqueness if we make the owner specification mandatory, otherwise, this requires a kind of central repository of hooks names, which might be complex to arrange

Reply:
Annotations are key/value pairs.  Each hook name needs to be unique because each key has to be unique.

So 'pre-drain.hook.machine.cluster-api.x-k8s.io/custom-drainer' can only appear once in the annotation map.  Cluster administrators are ultimately responsible for determining which hooks are present on their clusters.  Ideally, each controller or other hook owning implementation can be configurable as to which name it's using, with a sane default.  There is a finite length these names can be, as specified in another section of this document.

Reply:
Agreed, my major concern is to avoid name clash among different provider/controllers introducing new annotations without coordination, but I'm fine with a soft recommendation as well

Reply:
My thought was in the future, we can have an example that implements some nice-to-haves in an owning controller.  Things like specifying exactly what annotations you want to set on the machine via a config.  Specifying checking for lack of another annotation or annotations before proceeding to do your work via config.  So, by default I set the '...hook.machine.cluster.x-k8s.io/my-org-custom-drain' but that could be configurable to '...hook.machine.cluster.x-k8s.io/other-org-custom-drain' as the admin sees fit.  There's a lot of flexibility, and I think we'll work out some good patterns over time.

Reply:
`pre-drain.hook.machine.cluster-api.x-k8s.io/` is 45 characters, leaving 18 for the user. `pre-terminate...` is 49 / 14 for the user. Maybe we should omit the `.hook.machine` part?

Reply:
63 character limit applies to the portion after the '/' character (aka, the name portion).  The 'prefix' portion has an additional 253 character limit.  I linked to this in the risks/mitigations section.

See here: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set

Reply:
Ah, thanks for the clarification!

Reply:
Thanks @michaelgugino for the reply! I'm making a final pass through here to resolve comments that are noncontroversial. I believe we're all in violent agreement, and your WIP PR seems straightforward and without issue. Hoping to be able to get this merged ASAP.